doc/man3/flux_future_promise.adoc

```diff
@@ -0,0 +1,163 @@
+destructor _destroy_, to the future under an optional _name_.
+If the _name_ is NULL, the object is stored anonymously.
+`flux_future_aux_get()` retrieves an object by _name_.
+Anonymous objects cannot be retrieved.  All aux objects are destroyed
+when the future is destroyed.
+
+`flux_future_set_flux()` may be used to associate a Flux broker handle
+with a future.  The handle may be retrieved from within a promise using
+`flux_future_get_flux()` and used to set up message handlers that
+fulfill the promise in the same manner as described for reactor watchers.
+
+Futures may "contain" other futures, to arbitrary depth.  That is, a
+promise may create futures and use their continuations to complete the
+containing future in the same manner as reactor watchers and message
+handlers.
+
+
+REACTOR CONTEXTS
+----------------
+
+Internally, a future can operate in two reactor contexts.
+The promise may be called in either or both contexts, depending on
+which synchronization functions are called by the user.
+
+The main reactor context involves the reactor passed to `flux_future_create()`.
+This reactor is expected to be run or re-entered by the user after a
+continuation is set up with `flux_future_then()`, and can process the
+future's watchers in parallel with other watchers registered by the
+application.  The call to `flux_future_then()` triggers the promise
+callback in this context.
+
+Alternatively, an internal reactor is created when `flux_future_get()` or
+`flux_future_check()` are called before the future is complete.
+The separate reactor allows these functions to wait _only_ for the
+future's events, without allowing unrelated watchers registered in the
+main reactor to run, which might complicate the application's control flow.
+After the internal reactor is created, the promise is called in this context.
+When invoked via `flux_future_check()`, the internal reactor is run with
+the FLUX_REACTOR_NOWAIT flag so the future's watchers have a chance to
+complete the future if they can do so without blocking.  When invoked
+via `flux_future_get()`, the internal reactor runs until the future is
+completed.
+
+Since the promise may be called in either reactor context (at most once each),
+and is unaware of which context that is, it should take care when managing
+any context-specific state not to overwrite the state from a prior call.
+The ability to attach objects with destructors anonymously to the future
+with `flux_future_aux_set()` may be useful for managing the life cycle
+of reactor watchers and message handlers created by promises.
+
+
+MESSAGE HANDLERS
+----------------
+
+To allow message handlers to be registered in either reactor context,
+`flux_future_get_flux()` is context sensitive.  If called in the main
+reactor context, it directly returns the broker handle registered with
+`flux_future_set_flux()`.  If called in the internal reactor context,
+it returns a clone of that handle, obtained with `flux_clone()`, and
```

Comment:
`flux_clone(3)` is mentioned here -- just a reminder that we might want a manpage for it, if you've not already done it.